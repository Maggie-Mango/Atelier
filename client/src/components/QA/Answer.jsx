import React, {useState, useEffect} from 'react';

const Answer = ({answer}) => {

  //[helpfulness, setHelpfulness] = useState(answer.helpfulness);

  return (
    <div className='qa-answer'>
      <div className='qa-question'>A:
          <span className ='qa-tiny'> {answer.body}</span>
      </div>
      <div className='qa-answerby qa-tiny'>
        <div>by {answer.answerer_name}, {answer.date.slice(0,10)}</div>
        <div className='qa-divider'>|</div>
        <div className='qa-helpful'> Helpful?</div>
        <div className='qa-underscore qa-helpful'> Yes ({answer.helpfulness})</div>
        <div className='qa-divider'>|</div>
        <div className='qa-underscore qa-tiny'> Report</div>
      </div>
    </div>
  )
}

export default Answer