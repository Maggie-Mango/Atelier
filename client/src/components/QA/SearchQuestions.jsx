import React, {useState, useEffect} from 'react';

const SearchQuestions = (props) => {
  return (
    <div className='qa-search-bar'>
        <input type='text' />
        <button className="qa-search-btn">
          <span>Magifier</span>
        </button>

    </div>


  )
}


export default SearchQuestions;