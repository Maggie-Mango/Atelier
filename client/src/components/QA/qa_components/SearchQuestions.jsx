import React, {useState, useEffect} from 'react';

const SearchQuestions = ({onChange}) => {
  return (
    <div className='qa-search-bar'>
        <input
           type='text'
           placeholder='HAVE A QUESTION? SEARCH FOR ANSWER...'
           onChange={(e)=>onChange(e)}
        />
        {/* <button className="qa-search-btn">
          <span>Magifier</span>
        </button> */}
        <button className="qa-search-btn">
         <span className="glyphicon glyphicon-search"></span>
        </button>

    </div>


  )
}


export default SearchQuestions;