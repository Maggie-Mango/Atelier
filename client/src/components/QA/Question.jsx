import React, { useEffect, useState} from 'react'
import Answer from './Answer.jsx'

const Question = ({question}) => {

  // const [answerData, setAnswerData] = useState([]);
  // const [questionBody, setQuestionBody] = useState('');
  // const [questionHelpfulness, setQuestionHelpfulness] = useState();
  // const [answerData, setAnswerData] =useState();
  const [questionData, setQuestionData] = useState({
    questionBody: question.question_body,
    questionHelpfulness: question.question_helpfulness,
    answers: question.answers //object
  })
  const [visible, setVisible] = useState(2)
  const [loadMore, setLoadMore] = useState(false)

  const handleAddAnswer = () => {
    //do we need to post this?
    console.log('add answer')
  }

  const handleHelpful = () => {
    console.log('yes, helpful')
  }

  const {questionBody, questionHelpfulness, answers} = questionData;
  const checkAnswerLength = (({answers}) => {
    if (answers.length > 2) {
      setLoadMore(true)
    }
  })

  useEffect(()=> {
   if (Object.keys(answers).length>2) {
    setLoadMore(true)
   }
  }, []);


  return (
    <div>
      <div className='container'>
        <div className='question'>Q: {questionBody} </div>
        <div className='tiny helpful'>
          <div className= 'helpful'> Helpful?</div>
          <div className='underscore helpful' onClick={handleHelpful}> Yes ({questionHelpfulness})</div>
          <div className='divider'>|</div>
          <div className='underscore helpful' onClick={handleAddAnswer}> Add answers </div>
        </div>
      </div>
      <div className='answers'>
        {
            Object.keys(answers).slice(0, visible).map((key) => {
              return <Answer answer={answers[key]} />
            })
          }
      </div>

      {loadMore? <div className ='moreanswer tiny'>LOAD MORE ANSWERS</div> : null}
    </div>
  )

}

export default Question;
