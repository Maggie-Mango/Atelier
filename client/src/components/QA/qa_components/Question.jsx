import React, { useEffect, useState} from 'react'
import Answer from './Answer.jsx'
import axios from 'axios'
import AnswerModal from './AnswerModal.jsx'
import Helpful from './Helpful.jsx'

const Question = ({question, productId}) => {

  const [questionBody, setQuestionBody] = useState(question.question_body)
  const [answers, setAnswers] = useState(question.answers)

  const [visible, setVisible] = useState(2)
  const [loadMoreAnswers, setLoadMoreAnswers] = useState() //null, load more answers, collapse answers
  const [addAnswer, setAddAnswer] = useState(false)

  useEffect(() => {
   if (Object.keys(answers).length>2) {
    setLoadMoreAnswers('LOAD MORE ANSWERS')
   }
  }, []);

  const handleLoadMoreAnswers = (value) => {
    if (value === 'LOAD MORE ANSWERS') {
      setVisible(Object.keys(answers).length)
      setLoadMoreAnswers('COLLAPSE ANSWERS')
    } else if (value === 'COLLAPSE ANSWERS') {
      setVisible(2)
      setLoadMoreAnswers('LOAD MORE ANSWERS')
    }
  }


  const handleSubmitAnswer = () => {

    axios.get(`/qa/questions/${question.question_id}/answers`)
      .then((result) => setAnswers(result.data))//
      .catch((err) => console.log(err))

  }

  return (
    <div>
      <div className='qa-container'>
        <div className='qa-question'>Q: {questionBody} </div>
        <div className='qa-tiny qa-helpful'>
          <div className= 'qa-helpful'> Helpful?</div>
          <Helpful
            id={question.question_id}
            helpfulness={question.question_helpfulness}
            localStorageName={'questionsMarkedHelpful'}/>
          <div className='qa-divider'>|</div>
          <div className='qa-clickable qa-helpful' onClick={() => setAddAnswer(true)}> Add Answers </div>
        </div>
      </div>
      <div className='qa-answers qa-scroll'>
        <div className='qa-answers-left'>A: </div>
        <div className='qa-answers-right'>
          {Object.keys(answers).slice(0, visible).map((key) => {
                return <Answer answer={answers[key]} key={key} />
              })
          }
        </div>
      </div>

      {loadMoreAnswers ?
        <div className ='qa-moreanswer qa-tiny' onClick={() => handleLoadMoreAnswers(loadMoreAnswers)}>
          {loadMoreAnswers}</div>
      : null}

      <AnswerModal
        open={addAnswer}
        productName={'test'}
        question={questionBody}
        questionId={question.question_id}
        onClose={()=> {
          setAddAnswer(false)}}
        onSubmitAnswer={handleSubmitAnswer}
      />
    </div>
  )

}

export default Question;
