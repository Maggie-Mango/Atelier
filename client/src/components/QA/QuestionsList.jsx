import React from 'react'
import Question from './Question.jsx'

const QuestionsList = (props) => {
  return (
    <div className='main-questions'>
      {props.questions.map(question => <Question question={question} key={question.question_id}/>)}
    </div>
  )
}


export default QuestionsList;