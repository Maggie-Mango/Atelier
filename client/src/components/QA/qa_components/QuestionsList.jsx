import React, {useState, useEffect} from 'react'
import Question from './Question.jsx'
import QuestionModal from './QuestionModal.jsx';
import axios from 'axios';

const QuestionsList = ({questions, productId, productInfo}) => {

  const [questionDisplayCount, setQuestionDisplayCount] = useState(2);
  const [questionList, setQuestionList] = useState([]);
  const [moreQuestionVisible, setMoreQuestionVisible] = useState(false);
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    // console.log('initial render and when questions is updated ')
    if (questions) {
      setQuestionDisplayCount(2) //if new question list, then set it to display two
      if (questions.length>2) {
        setQuestionList(questions)
        setMoreQuestionVisible(true)
      } else {
        setQuestionList(questions)
        setMoreQuestionVisible(false)
      }
    }

  },[questions])

  useEffect(() => {
    // console.log('initial and rerender on display count')
    if (questionDisplayCount >= questions.length) {
      setMoreQuestionVisible(false)
    }
  },[questionDisplayCount])

  const handleSubmitQuestion = () => {

    axios.get(`/qa/questions/${productId}`)
      .then((result) => {//might add sort function here? new q has 0 helpful and will be add on the bottom
        setQuestionList(result.data)})
      .catch((err) => console.log(err))

  }

  return (
    <div>
      <div className='qa-main-questions'>
        {questionList.slice(0,questionDisplayCount).map(question =>
          <Question
            question={question}
            key={question.question_id}
            productId={productId}
            productName={productInfo? productInfo.name : 'product'}
            />)
        }
      </div>

      <div className='qa-bottom'>
        {moreQuestionVisible ?
          <div className='qa-more-question' onClick={() => setQuestionDisplayCount(questionDisplayCount + 2)}>
              MORE ANSWERED QUESTIONS
          </div>
          : null
        }

        <div className='qa-more-question' onClick={() => {setIsOpen(true)}}>
          ASK A QUESTION +
        </div>
        <QuestionModal
            open={isOpen}
            productName={productInfo? productInfo.name : 'product'}
            productId={productId}
            onClose={()=> {setIsOpen(false)}}
            onSubmitQuestion={handleSubmitQuestion}
        />
      </div>
    </div>

  )
}


export default QuestionsList;