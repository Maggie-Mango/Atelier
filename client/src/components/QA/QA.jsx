import React from 'react';
import QuestionsList from './QuestionsList.jsx';
import SearchQuestions from './SearchQuestions.jsx';
import MoreQuestions from './MoreQuestions.jsx';
import AskAQuestion from './AskAQuestion.jsx';

class QA extends React.Component {
  constructor(props) {
    super(props);
    this.handleAskAQuestion = this.handleAskAQuestion.bind(this)
  }

  handleAskAQuestion(){
    console.log('clicked ask as question')
  }

  handleMoreQuestions(){
    console.log('clicked handleMoreQuestions')
  }


  render() {
    return (
      <div className='QA'>
        <h5>QUESTIONS & ANSWERS</h5>
        <SearchQuestions/>
        <QuestionsList/>
        <div>
          <MoreQuestions handleMoreQuestions={this.handleMoreQuestions.bind(this)}/>
          <AskAQuestion handleAskAQuestion={this.handleAskAQuestion}/>
        </div>
    </div>
    )
  }

}

export default QA;