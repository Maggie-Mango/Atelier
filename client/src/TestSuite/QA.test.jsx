
import React from 'react';
import renderer from 'react-test-renderer';
import {render, screen, cleanup, fireEvent, waitForElement} from '@testing-library/react';
import '@testing-library/jest-dom';

import QA from '../components/QA/QA.jsx';
import SearchQuestions from '../components/QA/qa_components/SearchQuestions.jsx';
import QuestionsList from '../components/QA/qa_components/QuestionsList.jsx';
import Question from '../components/QA/qa_components/Question.jsx';
import questions from './questions.js'


afterEach(cleanup);

describe('QA Component', () => {
  it('should render QA component', () => {
    const {getByTestId} = render(<QA/>);
    const qaElement = getByTestId('QA-render');
    expect(qaElement).toBeInTheDocument();
  })
})



describe("Search function", () => {

  it("should render the search bar", () => {//{questions, productId, productInfo}
    const {queryByPlaceholderText} = render(<SearchQuestions />)
    const searchInput = queryByPlaceholderText('HAVE A QUESTION? SEARCH FOR ANSWER...');
    expect(searchInput).toBeTruthy;
  })

  it('should change when type in the search bar', () => {
    const {debug, queryByPlaceholderText} = render(<SearchQuestions />)
    const searchInput = queryByPlaceholderText('HAVE A QUESTION? SEARCH FOR ANSWER...');
    fireEvent.change(searchInput, {target: {value: "test"}})
    // debug();
    expect(searchInput.value).toBe("test")
  })

  // it('should updates list when input character length is >= 3', () => {
  //   const {queryByPlaceholderText} = render(<SearchQuestions />)
  //   const searchInput = queryByPlaceholderText('Have a question? Search for answersâ€¦');
  //   fireEvent.change(searchInput, {target: {value: "test"}})
  //   expect(searchInput.value).toBe("test")
  // })
})


describe("Question List", () => {

  // it("should render a list of questions", () => {
  //   const {debug, getByText} = render(<QuestionsList questions={questions} productId={'59553'} productInfo={{name: 'Camo Onesie'}}/>)
  //   const questionsElement = getByText(/ser/)
  //   expect(questionsElement.toBeTruthy)
  // })

  // it("should render two questions on initial render", () => {
  //   const {debug, getByText} = render(<QuestionsList questions={questions} productId={'59553'} productInfo={{name: 'Camo Onesie'}}/>)
  //   const questionsElement = getByText(/ser/)
  //   expect(questionsElement.toBeTruthy)
  // })

  // it("should render 'more  for each question", () => {
  //   const { queryByPlaceholderText} = render(<SearchQuestions />)
  //   expect(queryByPlaceholderText('HAVE A QUESTION? SEARCH FOR ANSWER...').toBeTruthy)
  // })

})

describe("Each Question", () => {

  beforeEach(() => {
    render(<Question question={questions[0]} productName={'product1'}/>)
  })

  let text = questions[0].question_body;
  let questionHelpfulness = questions[0].question_helpfulness;
  let answerList = Object.keys(questions[0].answers);

  it("should render the question body", () => {
    expect(screen.queryByText(`Q: ${text}`).toBeTruthy)
  })

  it("should render helpful count for the question", () => {
    expect(screen.queryByText(`${questionHelpfulness}`, {exact: false}).toBeTruthy)
  })


  it("should render 'Add Answer' link", () => {
    expect(screen.queryByText('Add Answers').toBeTruthy)
  })

  it("should render 'Add Answer' modal when click on the 'Add Answer' link", () => {
    const addAnswerLink = screen.queryByText('Add Answers')
    fireEvent.click(addAnswerLink)
    screen.debug()
    expect(screen.queryByText('Submit your Answer').toBeInTheDocument)
  })

  it("should render a list of answers for each question", () => {
    answer = answerList[0]
    const answerElementOne = screen.queryByText(//)
    expect(answerElementOne.toBeTruthy)
  })

  //integration test
  it("should increase the count of helpful when click 'Yes'", () => {

  })
})





describe("Each Answer", () => {


})



// describe("'More Answered Questions' button", () => {

//   it("should render 'More Answered Questions' when there are more than 2 questions ", () => {
//     const { queryByPlaceholderText} = render(<SearchQuestions />)
//     expect(queryByPlaceholderText('HAVE A QUESTION? SEARCH FOR ANSWER...').toBeTruthy)
//   })

//   it("should hind 'More Answered Questions' when there are less than 2 questions ", () => {
//     const { queryByPlaceholderText} = render(<SearchQuestions />)
//     expect(queryByPlaceholderText('HAVE A QUESTION? SEARCH FOR ANSWER...').toBeTruthy)
//   })

// })

// describe('Question Modal Window', () => {
//   it('should open ask a question modal on click', () => {
//     const mockOnClick = jest.fn()
//     const { getByTestId } = render(<AddReviewButton onClick={mockOnClick()}/>);
//     const clickIndicator = getByTestId('ClickIndicator')
//     fireEvent.click(clickIndicator)
//     expect(mockOnClick).toHaveBeenCalledTimes(1)
//   })

// })


// test('should match snapshot',  () => {
//   const tree = renderer.create(
//     <QA
//       questions={questions}
//       productId={'60205'}
//       productInfo={{'name': 'test name'}}
//     />).toJSON()
//   expect(tree).toMatchSnapshot();
// })