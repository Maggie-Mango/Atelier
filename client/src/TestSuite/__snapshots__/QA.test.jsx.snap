// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snaoshot 1`] = `
<div
  className="QA"
  data-testid="QA-render"
>
  <h5>
    QUESTIONS & ANSWERS
  </h5>
  <div
    className="qa-search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="Have a question? Search for answersâ€¦"
      type="text"
    />
    <button
      className="qa-search-btn"
    >
      <span>
        Magifier
      </span>
    </button>
  </div>
  <div>
    <div
      className="qa-main-questions"
    />
    <div>
      <div
        className="qa-more-question"
        onClick={[Function]}
      >
        ASK A QUESTION +
      </div>
    </div>
  </div>
</div>
`;

exports[`should match snapshot 1`] = `
<div
  className="QA"
  data-testid="QA-render"
>
  <h3>
    QUESTIONS & ANSWERS
  </h3>
  <div
    className="qa-search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="HAVE A QUESTION? SEARCH FOR ANSWER..."
      type="text"
    />
    <button
      className="qa-search-btn"
    >
      <span
        className="glyphicon glyphicon-search"
      />
    </button>
  </div>
  <div>
    <div
      className="qa-main-questions"
    />
    <div>
      <div
        className="qa-more-question"
        onClick={[Function]}
      >
        ASK A QUESTION +
      </div>
    </div>
  </div>
</div>
`;
